src/main/java/org/prebid/server/bidder/criteo/CriteoBidder.java

```diff
@@ -0,0 +1,294 @@
+package org.prebid.server.bidder.criteo;
+
+import com.iab.openrtb.request.App;
+import com.iab.openrtb.request.Banner;
+import com.iab.openrtb.request.BidRequest;
+import com.iab.openrtb.request.Device;
+import com.iab.openrtb.request.Format;
+import com.iab.openrtb.request.Imp;
+import com.iab.openrtb.request.Regs;
+import com.iab.openrtb.request.Site;
+import com.iab.openrtb.request.User;
+import com.iab.openrtb.response.Bid;
+import org.apache.commons.lang3.StringUtils;
+import org.prebid.server.bidder.Bidder;
+import org.prebid.server.bidder.criteo.model.CriteoGdprConsent;
+import org.prebid.server.bidder.criteo.model.CriteoPublisher;
+import org.prebid.server.bidder.criteo.model.CriteoRequest;
+import org.prebid.server.bidder.criteo.model.CriteoRequestSlot;
+import org.prebid.server.bidder.criteo.model.CriteoResponse;
+import org.prebid.server.bidder.criteo.model.CriteoResponseSlot;
+import org.prebid.server.bidder.criteo.model.CriteoUser;
+import org.prebid.server.bidder.model.BidderBid;
+import org.prebid.server.bidder.model.BidderError;
+import org.prebid.server.bidder.model.HttpCall;
+import org.prebid.server.bidder.model.HttpRequest;
+import org.prebid.server.bidder.model.Result;
+import org.prebid.server.exception.PreBidException;
+import org.prebid.server.identity.IdGenerator;
+import org.prebid.server.json.DecodeException;
+import org.prebid.server.json.EncodeException;
+import org.prebid.server.json.JacksonMapper;
+import org.prebid.server.proto.openrtb.ext.request.ExtImpCriteo;
+import org.prebid.server.proto.openrtb.ext.request.ExtRegs;
+import org.prebid.server.proto.openrtb.ext.request.ExtUser;
+import org.prebid.server.proto.openrtb.ext.response.BidType;
+import org.prebid.server.util.HttpUtil;
+import io.vertx.core.MultiMap;
+import io.vertx.core.http.HttpMethod;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class CriteoBidder implements Bidder<CriteoRequest> {
+
+    private static final String TEST_UUID = "e77e016a-091d-4cdb-a759-29d57a1ffa48";
+
+    private final String endpointUrl;
+    private final JacksonMapper jsonMapper;
+    private final IdGenerator idGenerator;
+
+    public CriteoBidder(String endpointUrl, JacksonMapper jsonMapper, IdGenerator idGenerator) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.jsonMapper = Objects.requireNonNull(jsonMapper);
+        this.idGenerator = Objects.requireNonNull(idGenerator);
+    }
+
+    @Override
+    public Result<List<HttpRequest<CriteoRequest>>> makeHttpRequests(BidRequest bidRequest) {
+        final CriteoRequest.CriteoRequestBuilder criteoRequestBuilder = CriteoRequest.builder()
```

Comment:
Sometimes we will return as an error in condition `allSlotsAreContainingSameNetworkId(criteoRequestSlots)`
So lets move this builder  creation to place where its actually needed

Reply:
Also building CriteoRequest can be separate method